pkg/validators/am0010_k8s_resource_names.go

```diff
@@ -0,0 +1,234 @@
+	if msg := isValidNamespaceName(mb.AddonMeta.TargetNamespace); !msg.IsEmpty() {
+		failures = append(failures, prefixedFailureMsg("targetNamespace", msg))
+	}
+
+	return failures
+}
+
+func validateAllNamespaces(mb types.MetaBundle) []failureMsg {
+	var result []failureMsg
+
+	if msgs := areValidNamespaceNames(mb.AddonMeta.Namespaces...); len(msgs) > 0 {
+		result = append(result, prefixedFailureMsg(
+			"namespaces", joinFailureMsgs(msgs...),
+		))
+	}
+
+	if msgs := areValidAnnotationNames(keys(mb.AddonMeta.NamespaceAnnotations)...); len(msgs) > 0 {
+		result = append(result, prefixedFailureMsg(
+			"namespaceAnnotations", joinFailureMsgs(msgs...),
+		))
+	}
+
+	if msgs := areValidLabelNames(keys(mb.AddonMeta.NamespaceLabels)...); len(msgs) > 0 {
+		result = append(result, prefixedFailureMsg(
+			"namespaceLabels", joinFailureMsgs(msgs...),
+		))
+	}
+
+	return result
+}
+
+func validateCommonAnnotations(mb types.MetaBundle) []failureMsg {
+	if mb.AddonMeta.CommonAnnotations == nil {
+		return []failureMsg{}
+	}
+
+	annotations := *mb.AddonMeta.CommonAnnotations
+
+	result := make([]failureMsg, 0, len(annotations))
+
+	if msgs := areValidAnnotationNames(keys(annotations)...); len(msgs) > 0 {
+		result = append(result, prefixedFailureMsg(
+			"commonAnnotations", joinFailureMsgs(msgs...),
+		))
+	}
+
+	return result
+}
+
+func validateCommonLabels(mb types.MetaBundle) []failureMsg {
+	if mb.AddonMeta.CommonLabels == nil {
+		return []failureMsg{}
+	}
+
+	labels := *mb.AddonMeta.CommonLabels
+
+	result := make([]failureMsg, 0, len(labels))
+
+	if msgs := areValidLabelNames(keys(labels)...); len(msgs) > 0 {
+		result = append(result, prefixedFailureMsg(
+			"commonLabels", joinFailureMsgs(msgs...),
+		))
+	}
+
+	return result
+}
+
+func areValidNamespaceNames(nsNames ...string) []failureMsg {
+	msgs := make([]failureMsg, 0, len(nsNames))
+
+	for _, name := range nsNames {
+		if msg := isValidNamespaceName(name); !msg.IsEmpty() {
+			msgs = append(msgs, msg)
+		}
+	}
+
+	return msgs
+}
+
+func isValidNamespaceName(nsName string) failureMsg {
+	if valid, failureReasons := reasonsToResult(validation.ValidateNamespaceName(nsName, false)); !valid {
+		return prefixedFailureMsg(
+			fmt.Sprintf("\"%s\" is not a valid kubernetes namespace name", nsName),
+			failureMsg(failureReasons),
+		)
+	}
+
+	return emptyMsg
+}
+
+func areValidAnnotationNames(aNames ...string) []failureMsg {
+	msgs := make([]failureMsg, 0, len(aNames))
+
+	for _, name := range aNames {
+		if msg := isValidAnnotationName(name); !msg.IsEmpty() {
+			msgs = append(msgs, msg)
+		}
+	}
+
+	return msgs
+}
+
+func isValidAnnotationName(aName string) failureMsg {
+	if valid, failureReasons := reasonsToResult(utilvalidation.IsQualifiedName(strings.ToLower(aName))); !valid {
+		return prefixedFailureMsg(
+			fmt.Sprintf("\"%s\" is not a valid kubernetes annotation name", aName),
+			failureMsg(failureReasons),
+		)
+	}
+
+	return emptyMsg
+}
```

Comment:
nit: Is same as `isValidLabelName`, except the message and `strings.ToLower`. Is this intended?

Reply:
Yes, the `strings.ToLower` part is the only difference as annotations allow mixed-case prefixes. As for refactoring I could write a wrapper around `IsQualifiedName` that accepts a resource type, but it's kind of trivial. I can play around with it though.